### ISSUE TITLE
[REVIEW]: Opfi: A Python package for identifying gene clusters in large genomics and metagenomics data sets

### ISSUE THREAD

**Issue by whedon** [openjournals/joss-reviews#3678]
**Submitting author:** @alexismhill3 (<a href="http://orcid.org/0000-0002-5382-8851">Alexis Hill</a>)
**Repository:** <a href="https://github.com/wilkelab/Opfi" target ="_blank">https://github.com/wilkelab/Opfi</a>
**Version:** 0.1.0
**Editor:** @csoneson
**Reviewer:** @Thomieh73, @afrubin
**Archive:** Pending

**:warning: JOSS reduced service mode :warning:**

Due to the challenges of the COVID-19 pandemic, JOSS is currently operating in a "reduced service mode". You can read more about what that means in [our blog post](https://blog.joss.theoj.org/2020/05/reopening-joss).

## Status

[![status](https://joss.theoj.org/papers/cc0a3b03e9cff97e5f93e1a3a8a1808d/status.svg)](https://joss.theoj.org/papers/cc0a3b03e9cff97e5f93e1a3a8a1808d)

Status badge code:

```
HTML: <a href="https://joss.theoj.org/papers/cc0a3b03e9cff97e5f93e1a3a8a1808d"><img src="https://joss.theoj.org/papers/cc0a3b03e9cff97e5f93e1a3a8a1808d/status.svg"></a>
Markdown: [![status](https://joss.theoj.org/papers/cc0a3b03e9cff97e5f93e1a3a8a1808d/status.svg)](https://joss.theoj.org/papers/cc0a3b03e9cff97e5f93e1a3a8a1808d)
```
**Reviewers and authors**:

Please avoid lengthy details of difficulties in the review thread. Instead, please create a new issue in the <a href="https://github.com/wilkelab/Opfi" target="_blank">target repository</a> and link to those issues (especially acceptance-blockers) by leaving comments in the review thread below. (For completists: if the target issue tracker is also on GitHub, linking the review thread in the issue or vice versa will create corresponding breadcrumb trails in the link target.)

## Reviewer instructions & questions

@Thomieh73 & @afrubin, please carry out your review in this issue by updating the checklist below. If you cannot edit the checklist please:
1. Make sure you're logged in to your GitHub account
2. Be sure to accept the invite at this URL: https://github.com/openjournals/joss-reviews/invitations

The reviewer guidelines are available here: https://joss.readthedocs.io/en/latest/reviewer_guidelines.html. Any questions/concerns please let @csoneson know.

✨ **Please start on your review when you are able, and be sure to complete your review in the next six weeks, at the very latest** ✨ 

## Review checklist for @Thomieh73
✨ Important: Please do not use the _Convert to issue_ functionality when working through this checklist, instead, please open any new issues associated with your review <a target="_blank" href="https://github.com/wilkelab/Opfi">in the software repository associated with the submission</a>. ✨

### Conflict of interest

- [ ] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [ ] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [ ] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/wilkelab/Opfi">repository url</a>?
- [ ] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [ ] **Contribution and authorship:** Has the submitting author (@alexismhill3) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?
- [ ] **Substantial scholarly effort:** Does this submission meet the scope eligibility described in the [JOSS guidelines](https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort)

### Functionality

- [ ] **Installation:** Does installation proceed as outlined in the documentation?
- [ ] **Functionality:** Have the functional claims of the software been confirmed?
- [ ] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [ ] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [ ] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [ ] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [ ] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [ ] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [ ] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [ ] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [ ] **A statement of need:** Does the paper have a section titled 'Statement of Need' that clearly states what problems the software is designed to solve and who the target audience is?
- [ ] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [ ] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [ ] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

## Review checklist for @afrubin
✨ Important: Please do not use the _Convert to issue_ functionality when working through this checklist, instead, please open any new issues associated with your review <a target="_blank" href="https://github.com/wilkelab/Opfi">in the software repository associated with the submission</a>. ✨

### Conflict of interest

- [ ] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [ ] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [ ] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/wilkelab/Opfi">repository url</a>?
- [ ] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [ ] **Contribution and authorship:** Has the submitting author (@alexismhill3) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?
- [ ] **Substantial scholarly effort:** Does this submission meet the scope eligibility described in the [JOSS guidelines](https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort)

### Functionality

- [ ] **Installation:** Does installation proceed as outlined in the documentation?
- [ ] **Functionality:** Have the functional claims of the software been confirmed?
- [ ] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [ ] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [ ] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [ ] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [ ] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [ ] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [ ] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [ ] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [ ] **A statement of need:** Does the paper have a section titled 'Statement of Need' that clearly states what problems the software is designed to solve and who the target audience is?
- [ ] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [ ] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [ ] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

**Comment by whedon** [comment]
Hello human, I'm @whedon, a robot that can help you with some common editorial tasks. @Thomieh73, @afrubin it looks like you're currently assigned to review this paper :tada:.

**:warning: JOSS reduced service mode :warning:**

Due to the challenges of the COVID-19 pandemic, JOSS is currently operating in a "reduced service mode". You can read more about what that means in [our blog post](https://blog.joss.theoj.org/2020/05/reopening-joss).

**:star: Important :star:**

If you haven't already, you should seriously consider unsubscribing from GitHub notifications for this (https://github.com/openjournals/joss-reviews) repository. As a reviewer, you're probably currently watching this repository which means for GitHub's default behaviour you will receive notifications (emails) for all reviews 😿

To fix this do the following two things:

1. Set yourself as 'Not watching' https://github.com/openjournals/joss-reviews:

![watching](https://cloud.githubusercontent.com/assets/4483/20250593/64d7ce48-a9de-11e6-9225-d3dfb3e48e68.png)

2. You may also like to change your default settings for this watching repositories in your GitHub profile here: https://github.com/settings/notifications

![notifications](https://cloud.githubusercontent.com/assets/4483/20250604/79eb3144-a9de-11e6-9e51-0fe940966689.png)


For a list of things I can do to help you, just type:

```
@whedon commands
```

For example, to regenerate the paper pdf after making changes in the paper's md or bib files, type:

```
@whedon generate pdf
```

**Comment by whedon** [comment]
Wordcount for `paper.md` is 1401

**Comment by whedon** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.1186/2042-5783-2-3 is OK
- 10.1038/nrg1709 is OK
- 10.1186/1471-2105-10-421 is OK
- 10.1038/nbt.3988 is OK
- 10.1038/s41592-021-01101-x is OK
- 10.1093/nar/gkz310 is OK
- 10.1093/nar/gky383 is OK
- 10.1093/nar/gkz192 is OK
- 10.1186/1471-2105-8-18 is OK
- 10.1093/bioinformatics/btaa213 is OK
- 10.1101/2021.08.16.456562 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by whedon** [comment]
```
Software report (experimental):

github.com/AlDanial/cloc v 1.88  T=0.31 s (525.7 files/s, 84309.5 lines/s)
--------------------------------------------------------------------------------
Language                      files          blank        comment           code
--------------------------------------------------------------------------------
C++                              85           2266            585          12905
Python                           38           1095           1366           4350
C/C++ Header                     19            261             61           1418
Markdown                          5            109              0            305
XML                               1              0              0            210
TeX                               1             10              0            179
reStructuredText                  8            202            244            177
YAML                              3             11             12            103
Jupyter Notebook                  1              0            302             77
make                              2             12              8             28
Bourne Again Shell                1              1              0              4
--------------------------------------------------------------------------------
SUM:                            164           3967           2578          19756
--------------------------------------------------------------------------------


Statistical information for the repository 'fbe5dc8158f92bfb3e7b73b1' was
gathered on 2021/08/31.
The following historical commit information, by author, was found:

Author                     Commits    Insertions      Deletions    % of changes
Alexis Hill                    146         23336           3224           72.53
Jim Rybarski                   159          7274           2786           27.47

Below are the number of rows from each author that have survived and are still
intact in the current revision:

Author                     Rows      Stability          Age       % in comments
Alexis Hill               19831           85.0         17.4                5.03
Jim Rybarski               4476           61.5         12.2                8.80
```

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.03678/joss.03678/10.21105.joss.03678.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.03678/joss.03678/10.21105.joss.03678.pdf) :page_facing_up: :point_left:

**Comment by csoneson** [comment]
👋🏼 @alexismhill3, @clauswilke, @Thomieh73, @afrubin - this is the review thread for the submission. All of our communications will happen here from now on.

Both reviewers have checklists at the top of this thread. As you go over the submission, please check any items that you feel have been satisfied. There are also links to the JOSS reviewer guidelines.

The JOSS review is different from most other journals. Our goal is to work with the authors to help them meet our criteria instead of merely passing judgment on the submission. As such, the reviewers are encouraged to submit issues directly in the software repository. If you do so, please mention this thread so that a link is created (and I can keep an eye on what is happening). Please also feel free to comment and ask questions in this thread. It is often easier to post comments/questions/suggestions as you come across them instead of waiting until you've reviewed the entire package.

Please feel free to ping me (@csoneson) if you have any questions or concerns. Thanks!

**Comment by afrubin** [comment]
@csoneson not sure why, but I'm unable to check any of the boxes in my checklist.

**Comment by afrubin** [comment]
@alexismhill3 there are several (mostly old) open issues in the GitHub repository at this time. What is the status of these issues? Some of them seem worth fixing, especially the temp file (https://github.com/wilkelab/Opfi/issues/181) and missing documentation (https://github.com/wilkelab/Opfi/issues/30) issues.

It would be helpful to know before we start adding our own comments during the review.

**Comment by clauswilke** [comment]
@afrubin The vast majority of the old issues were stale and we have started closing them. The temp-file issue unfortunately is caused by a dependency and doesn't have an easy fix. Documentation has improved massively since spring 2020.

Any requests for improvement, new features, or bug fixes should be filed as new issues. Thanks!

**Comment by Thomieh73** [comment]
@csoneson Hi, I am having issues with notifications from this repo. I am receiving all notifications from this entire repo. I have pushed the button to unsubscribe from this thread, but somehow I am stil receiving all notifications from this repo. Any suggestions on how to solve this.
It drives me nuts to have 100 mails in my mailbox everyday that are not relevant.

**Comment by csoneson** [comment]
@Thomieh73 - check the instructions in the second post above: https://github.com/openjournals/joss-reviews/issues/3678#issuecomment-909628355

**Comment by Thomieh73** [comment]
@csoneson Thanks I had completely forgotten about that option at github. :-)

**Comment by whedon** [comment]
:wave: @Thomieh73, please update us on how your review is going (this is an automated reminder).

**Comment by whedon** [comment]
:wave: @afrubin, please update us on how your review is going (this is an automated reminder).

**Comment by afrubin** [comment]
I started with the installation instructions and have opened some related issues:
https://github.com/wilkelab/Opfi/issues/198
https://github.com/wilkelab/Opfi/issues/199
https://github.com/wilkelab/Opfi/issues/200
https://github.com/wilkelab/Opfi/issues/201

**Comment by alexismhill3** [comment]
Thanks @afrubin, I'll get started working on these

**Comment by Thomieh73** [comment]
Hi @alexismhill3,
I had created two issues when working on the installation and on your tutorial (which you already solved).
https://github.com/wilkelab/Opfi/issues/202 - which is a suggestion of how one can create a conda environment with all the dependencies.

I did not find any other problems using the tool or checking the documentation

**Comment by Thomieh73** [comment]
@csoneson I am done with the checklist. It is an interesting tool due to the versatility of what you can research.

**Comment by alexismhill3** [comment]
Awesome, thanks @Thomieh73! 
I'm working on creating a conda package recipe for Opfi, per your suggestion, which I agree is a good idea. I think this will also resolve some other issues raised by @afrubin about the clarity of the installation process, since it will basically be condensed down to `conda install opfi`

**Comment by afrubin** [comment]
@alexismhill3 I'm waiting to try out the new installation method, but in the meantime here are a few comments on the manuscript.

I was a little confused by the exclusive focus on metagenomics in the statement of need section. This led me to expect that the program would be analyzing some sort of shotgun metagenomic data, rather than a single genome assembly. Clearly many new genomes are arising from metagenomic studies, but this could be clearer from the start.

The implementation says that Opfi is implemented entirely in Python, but the repository does contain an optional C dependency. This should be mentioned here.

The use of capitalization in the manuscript is inconsistent (e.g. Cas1 vs. cas7, "Blastp" which is usually stylized as blastp or BLASTP). There are other formatting inconsistencies, for example Rippkea orientalis is not italicized. The authors should perform a detailed pass through the manuscript to address these kind of issues.

**Comment by afrubin** [comment]
I have added a couple more GitHub issues arising from the manuscript:
https://github.com/wilkelab/Opfi/issues/205
https://github.com/wilkelab/Opfi/issues/206

**Comment by alexismhill3** [comment]
@afrubin I've created a conda recipe for Opfi and updated the documentation accordingly. These changes were merged, so you should be able to see them on the main branch now. 

I had to update a few unit tests during this process. Specifically, some tests required the modified version of PILER-CR, which we have decided to deprecate the use of (see wilkelab/Opfi#207 for more details). So if you were intending to run the unit test suite, you should grab the new changes before doing so.

**Comment by clauswilke** [comment]
In brief, we have made all changes except the API changes requested in issues https://github.com/wilkelab/Opfi/issues/205 and https://github.com/wilkelab/Opfi/issues/206. We agree with these requests in principle, but they are breaking changes and that would create major problems for us in the code for a related paper (https://www.biorxiv.org/content/10.1101/2021.08.16.456562v2). This paper is about to formally appear, and we don't want to end up with a situation where this client code is immediately incompatible with the latest official release of Opfi.

We will leave these issues open and may address them at a future point.

Regarding the term "Operon" (https://github.com/wilkelab/Opfi/issues/205), we have made changes to the documentation to make it absolutely clear that we are not claiming that an `Operon` object in the library is definitely an operon in the biological meaning.

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.03678/joss.03678/10.21105.joss.03678.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.03678/joss.03678/10.21105.joss.03678.pdf) :page_facing_up: :point_left:

**Comment by afrubin** [comment]
@alexismhill3 @clauswilke I understand the tension between wanting to update but avoiding breaking changes (especially with another manuscript involved), and I'm satisfied with the way you addressed the outstanding issues. I hope this software is widely-adopted and that you have the time and motivation to come back to them!

My comments on the manuscript have been addressed in full. The only outstanding issue is that you should cite Bioconda in the manuscript: https://bioconda.github.io/#citing-bioconda

FYI I was able to install Opfi and run the tests using Conda via Docker: https://docs.anaconda.com/anaconda/user-guide/tasks/docker/

I don't think you need to add this to the documentation (unless you want to support this use case) but I wanted to mention it in case it's useful for you or another user later on.

**Comment by afrubin** [comment]
@csoneson Yes, I am happy with the submission (once the Bioconda reference is added).

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.03678/joss.03678/10.21105.joss.03678.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.03678/joss.03678/10.21105.joss.03678.pdf) :page_facing_up: :point_left:

**Comment by alexismhill3** [comment]
Thanks everyone! @csoneson @afrubin I've added the Bioconda reference.
